### PR TITLE
Generate installation script and Homebrew formula

### PR DESCRIPTION
**Pull request by Duhemm**
I rewrote the installation script in Python and tested it on Windows. It appears to be working.

I also added a Homebrew formula that uses the Python script to do the installation. I haven't been able to completely test it at the moment because we don't publish zinc and nailgun, so they're not found during installation (homebrew builds packages in a sandbox).

The scripts are generated with a very simple thing. Running `frontend/makeTemplates` generates the scripts.

Finally, I added a script to create releases in `bin/make-tag.sh`. Example usage: `bin/make-tag.sh v0.5`, `bin/make-tag.sh v0.5 my-remote-name`.

It'll create tags in both the Homebrew formula and Bloop's repo and push all of that. The remote name can be specified.

I know that's a lot of heterogeneous looking stuff, but everything was kind of built together...

### CODE REVIEW

**Comment by olafurpg** [comment]
Let me know when you want a dummy tester for the new installation setup on mac!

**Comment by Duhemm** [comment]
I will, thanks @olafurpg !

File: build.sbt [hunk 1]
```diff
@@ -129,18 +130,21 @@ val bloop = project
...
+    s"^${sbtBloop.id}/$publishLocalCmd",
+    s"${mavenBloop.id}/$publishLocalCmd"
+  ).mkString(";", ";", "")
+)
 
-addCommandAlias("runTests", List(
-  s"${sbtBloop.id}/${scriptedAddSbtBloop.key.label}",
-  s"${sbtBloop.id}/${scripted.key.label}"
-).mkString(";", ";", ""))
+addCommandAlias("runTests",
```
Comment (jvican) [inline build.sbt]: Can we format this one like `setupIntegrations` :smile:?

File: frontend/src/main/scala/bloop/Cli.scala [hunk 2]
```diff
@@ -23,10 +23,16 @@ object Cli {
...
     val command = ngContext.getCommand
-    val args =
+    val args = {
       if (command == "bloop.Cli") ngContext.getArgs
       else command +: ngContext.getArgs
-    val cmd = parse(args, nailgunOptions)
+    }
+    val cmd = {
+      // If no command is given to bloop, we'll receive the script's name.
+      if (command == "bloop")
```
Comment (jvican) [inline frontend/src/main/scala/bloop/Cli.scala]: Why is this necessary?

File: project/BuildPlugin.scala [hunk 3]
```diff
@@ -53,17 +55,26 @@ object BuildKeys {
...
-    integrationTestsLocation := (Keys.baseDirectory in sbt.ThisBuild).value / "integration-tests" / "integration-projects",
+    integrationTestsLocation := buildBase.value / "integration-tests" / "integration-projects",
     Keys.testOptions += Tests.Argument(TestFrameworks.JUnit, "-v", "-a"),
     Keys.libraryDependencies ++= List(
       Dependencies.junit % Test
     ),
   )
 
+  import ohnosequences.sbt.GithubRelease.{keys => GHReleaseKeys}
+  val releaseSettings = Seq(
```
Comment (jvican) [inline project/BuildPlugin.scala]: So beautiful 💛

File: project/ReleaseUtils.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
+package build
+
+import java.io.File
+
+import sbt.{Def, Keys, MessageOnlyException}
+import sbt.io.syntax.fileToRichFile
+import sbt.io.IO
+
+/** Utilities that are useful for releasing Bloop */
+object ReleaseUtils {
```
Comment (jvican) [inline project/ReleaseUtils.scala]: This looks awesome, great work 👍

File: project/ReleaseUtils.scala [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  /**
+   * Creates a new installation script (based on the normal installation script) that has default
+   * values for the nailgun commit and version of Bloop to install.
+   *
+   * This lets us create an installation script that doesn't need any additional input to install
+   * the version of Bloop that we're releasing.
+   */
+  val versionedInstallScript = Def.task {
+    IO.readLines(installScript.value) match {
+      case shebang :: rest =>
```
Comment (jvican) [inline project/ReleaseUtils.scala]: Dynamic scripts FTW

File: project/ReleaseUtils.scala [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+/** Utilities that are useful for releasing Bloop */
+object ReleaseUtils {
+
+  /** The path to our installation script */
+  private val installScript = Def.setting { BuildKeys.buildBase.value / "bin" / "install.py" }
+
+  /** The latest commit in the `nailgun` submodule. */
+  private val nailgunCommit = Def.setting {
+    GitUtils.withGit(GitUtils.submoduleGitDir("nailgun").value) { git =>
```
Comment (jvican) [inline project/ReleaseUtils.scala]: A small question: why is this necessary? Can we not reuse sbt dynver's utils for nailgun? `version in NailgunBuild` should give us this commit hash 🤔

File: project/ReleaseUtils.scala [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  /**
+   * The content of the Homebrew Formula to install the version of Bloop that we're releasing.
+   *
+   * @param version The version of Bloop that we're releasing (usually `Keys.version.value`)
+   * @param tagName The name of the tag that we're releasing
+   * @param installSha The SHA-256 of the versioned installation script.
+   */
+  def formulaContent(version: String, tagName: String, installSha: String): String = {
```
Comment (jvican) [inline project/ReleaseUtils.scala]: I'm really happy with how everything has turned out to be

File: project/build.sbt [hunk 8]
```diff
@@ -4,6 +4,8 @@ val root = project
   .in(file("."))
   .settings(
     scalaVersion := "2.12.4",
+    resolvers += "Jenkins repo" at "http://repo.jenkins-ci.org/public/",
```
Comment (jvican) [inline project/build.sbt]: Agrrrrr.

File: project/BuildPlugin.scala [hunk 9]
```diff
@@ -53,17 +55,26 @@ object BuildKeys {
...
+    GHReleaseKeys.ghreleaseRepoOrg := "scalacenter",
+    GHReleaseKeys.ghreleaseRepoName := "bloop",
+    GHReleaseKeys.ghreleaseAssets += ReleaseUtils.versionedInstallScript.value,
```
Comment (jvican) [inline project/BuildPlugin.scala]: This means we're running `versionedInstallScript` twice, one here and one in `updateHomebrewFormula`. Would it be possible to make `versionedInstallScript` a setting instead of a task?

File: build.sbt (hunk 1 above)
Comment (Duhemm) [reply build.sbt]: scalafmt doesn't agree -- do you really want to go against it?

Comment (olafurpg) [reply build.sbt]: @Duhemm http://scalameta.org/scalafmt/#Configstyle

Comment (olafurpg) [reply build.sbt]:
```
align = none
danglingParentheses = true
```
gives you this by default, fwiw

File: project/ReleaseUtils.scala (hunk 6 above)
Comment (Duhemm) [reply project/ReleaseUtils.scala]:
```scala
scala> (version in build.BuildKeys.NailgunBuild).eval
res0: String = 0.1-SNAPSHOT
```

I didn't manage to get the value I wanted with `gitHeadCommit` either...

File: frontend/src/main/scala/bloop/Cli.scala (hunk 2 above)
Comment (Duhemm) [reply frontend/src/main/scala/bloop/Cli.scala]: It's not necessary for this PR, I didn't intend to keep this commit here...
The reason is that at the moment, calling `bloop` without any argument will print `bloop: Command not found` which is extraordinarily surprising for people trying bloop for the first time. I'm happy to remove it and put that in another PR (as I initially intended to) if you prefer.

File: build.sbt (hunk 1 above)
Comment (Duhemm) [reply build.sbt]: Actually scalafmt is smart enough to understand what I want to do on its own 😱

File: project/BuildPlugin.scala (hunk 9 above)
Comment (Duhemm) [reply project/BuildPlugin.scala]: I changed it, but now it's a side-effecting setting which is not super cool either...

Comment (jvican) [reply project/BuildPlugin.scala]: Oh damn... if it's side-effecting, it's better as a task. Can we have it as a cache that you only run once (with some kind of semaphore or sync utility) please? 🙏

File: project/ReleaseUtils.scala (hunk 6 above)
Comment (jvican) [reply project/ReleaseUtils.scala]: We can keep it this way and I'll make sbt-dynver work and remove this piece of code in the future.

File: frontend/src/main/scala/bloop/Cli.scala (hunk 2 above)
Comment (jvican) [reply frontend/src/main/scala/bloop/Cli.scala]: > It's not necessary for this PR, I didn't intend to keep this commit here...
The reason is that at the moment, calling bloop without any argument will print bloop: Command not found which is extraordinarily surprising for people trying bloop for the first time. I'm happy to remove it and put that in another PR (as I initially intended to) if you prefer.

Comment (jvican) [reply frontend/src/main/scala/bloop/Cli.scala]: Does this happen even with our latest changes in nailgun?
